### PR TITLE
Require go 1.18

### PR DESCRIPTION
**Pull request by MichaelEischer**
What does this PR change? What problem does it solve?
-----------------------------------------------------
#4038 and #4028 require Go 1.18 to build. This rather large bump in the required minimum version is a result of making use of the generics support which was introduced in Go 1.18.

This PR also cleans up some no longer necessary workarounds and build constraints.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
See #4028.

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- ~~[ ] I have added tests for all code changes.~~
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

**Comment by greatroar** [comment]
github.com/hashicorp/golang-lru/v2 would also require 1.18.

**Comment by greatroar** [comment]
Code changes LGTM. I suppose this (or any upgrade to >=1.16) would also close #2659.

File: changelog/unreleased/pull-4041 [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Change: Update dependencies and require Go 1.18 or newer
+
+We've updated most dependencies. Since some libraries require newer language
+features we're dropping support for Go 1.15 - 1.17, which means that restic now
```
Comment (greatroar) [inline changelog/unreleased/pull-4041:4]:
```suggestion
features, we're dropping support for Go 1.15 - 1.17, which means that restic now
```

We're not dropping language features :)

Comment (MichaelEischer) [reply changelog/unreleased/pull-4041]: Thanks :-)

**Comment by MichaelEischer** [comment]
LGTM.

**Pull request merged**